### DISCUSSION TITLE
Question about attributes and impl_class - generic_relationship in sqla-utils

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#11044] [Usage Questions] [answered]
I was trying to upgrade my sqlalchemy from 2.0.20 to 2.0.27
And came across this issue with sqlalchemy-utils which affected me:
https://github.com/kvesteri/sqlalchemy-utils/pull/725

There was a comment about whether that was an issue in sqlalchemy, and I went digging and found:
https://github.com/sqlalchemy/sqlalchemy/blob/rel_2_0_27/lib/sqlalchemy/orm/attributes.py#L2610

(Note that I do not really understand a lot of this (attributes, impl_class, etc.) but I followed the code 🚴)

It looks like in this area, the arguments are different when `impl_class` is provided vs not provided.
```py
if impl_class:
    impl = impl_class(
        class_, key, dispatch, **kw     # Earlier: class_, key, typecallable, dispatch, **kw
    )
else ...
    impl = CollectionAttributeImpl(
        class_, key, callable_, dispatch, typecallable=typecallable, **kw
    )
```

I'm wondering if sqla-utils is doing an anti-pattern of inheriting from `ScalarAttributeImpl` or if this is something that should be fixed on sqlalchemy side

It feels like this was a purposeful decision in:
https://github.com/sqlalchemy/sqlalchemy/commit/0a12d22d6d63bae1c695f0afc90f9cb6d825b82e#diff-049ddc7ced64ddad4975893bfdb4c39a6e3492d3cfcd2429bbfbc3e0f9c72535R166-R173

**Comment by zzzeek** [comment, marked as answer]
sqlalchemy-utils should not be inheriting from those impl classes is the basic answer, these are not designed for subclassing.

That said, if sqlalchemy-utils had a real need to be able to make a sqlalchemy-ish descriptor of some kind as they are doing there (I haven't looked at what it does), in some other universe they'd come to us, say "this is what we are trying to do and being able to subclass `QueryableAttributeImpl` would be just perfect", we would be potentially convinced to add an expansion point like a `UserDefinedAttributeImpl` or something where we'd make it have a fixed API (more likely we'd just [add additional event hooks](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.AttributeEvents), since subclassing is a very thorny endeavor), we'd add tests for it, and then as we go forward making changes to our own `AttributeImpl` etc. we'd make sure `UserDefinedWhatever` never changes.

in the current universe, everything in sqlalchemy-utils, while it's great that it's out there and its designers have provided code for people to use, none of it was done with any input from us whatsoever and I can't recall any time they came to us asking for any changes at all (I could be wrong, but if I am, it would have been 15 years ago) so anytime they needed a new hook to SQLAlchemy's internals, they just went and used the internals directly as we see here.

put another way, I've probably written (younger me variations of) this comment half a dozen times or more already over the years w.r.t. sqlalchemy-utils :)

**Reply by AbdealiLoKo** [reply]
Aaah, I see. This is good context - thanks for repeating it :)
